File: shovelling/shovelling.py
"""
https://open.kattis.com/problems/shovelling
"""

import sys
from typing import List
from queue import Queue

# Constants
BLOCKED = "#"
SNOW = "o"
CLEAR = "."
SHOVELLED = "~"


class Pos:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y


class Rect:
    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height


class Square:
    def __init__(self, pos: Pos, sign: str):
        self._pos = pos
        self._char = sign

    def getChar(self):
        return self._char

    def getNeighbours(self, size: Rect) -> List[Pos]:
        if self._pos.x > 0:
            yield Pos(self._pos.x - 1, self._pos.y)
        if self._pos.x < size.width - 1:
            yield Pos(self._pos.x + 1, self._pos.y)
        if self._pos.y > 0:
            yield Pos(self._pos.x, self._pos.y - 1)
        if self._pos.y < size.height - 1:
            yield Pos(self._pos.x, self._pos.y + 1)


class MapData:
    def __init__(self, size: Rect, data: str):
        self._size = size
        self._squares: List[Square] = [None] * len(data)
        self._dist: List[List[int]] = [None] * len(data)
        for y in range(size.height):
            for x in range(size.width):
                p = y*size.width + x
                self._squares[p] = Square(Pos(x, y), data[p])
        self.house_A = data.find('A')
        self.house_B = data.find('B')
        self.house_C = data.find('C')
        self.house_D = data.find('D')

    def printMap(self):
        print("{} {}".format(self._size.width, self._size.height))
        for y in range(self._size.height):
            line = ""
            for x in range(self._size.width):
                pos = y * self._size.width + x
                sq = self._squares[pos]
                line += str(sq.getChar())
            print(line)

    def getSquare(self, pos: Pos):
        return self._squares[pos.y*self._size.width + pos.x]

    def solve(self):
        pivot = self.getPivot()
        while pivot >= 0:
            self._squares[pivot]._char = SHOVELLED
            pivot = self.getPivot()

    def getPivot(self) -> int:
        self.calcDist(self.house_A)
        self.calcDist(self.house_B)
        self.calcDist(self.house_C)
        self.calcDist(self.house_D)
        d = [0] * len(self._squares)
        for i in range(len(self._squares)):
            d[i] += self._dist[self.house_A][i]
            d[i] += self._dist[self.house_B][i]
            d[i] += self._dist[self.house_C][i]
            d[i] += self._dist[self.house_D][i]
        minCostPivot = []
        minCostValue = 9999
        for i in range(len(d)):
            if d[i] == 0:
                return -1
            if self._squares[i].getChar() != SNOW:
                continue
            if d[i] < minCostValue:
                minCostPivot.clear()
                minCostPivot.append(i)
                minCostValue = d[i]
            elif d[i] == minCostValue:
                minCostPivot.append(i)
        return minCostPivot[0]

    def calcDist(self, start):
        dist = [9999] * len(self._squares)
        q = Queue()
        dist[start] = 0
        q.put(start)
        while not q.empty():
            next_pos = q.get()
            next: Square = self._squares[next_pos]
            for neighbour in next.getNeighbours(self._size):
                neighbour_pos = neighbour.y*self._size.width + neighbour.x
                neighbour_square = self._squares[neighbour_pos]
                if neighbour_square.getChar() == BLOCKED:
                    continue
                neighbour_cost = dist[next_pos]
                if neighbour_square.getChar() == SNOW:
                    neighbour_cost += 1
                if neighbour_cost < dist[neighbour_pos]:
                    dist[neighbour_pos] = neighbour_cost
                    q.put(neighbour_pos)
        self._dist[start] = dist


def readMaps():
    while True:
        dim = sys.stdin.readline().split()
        size = Rect(int(dim[0]), int(dim[1]))
        if size.width == 0 and size.height == 0:
            break

        map = ""
        for i in range(size.height):
            map += sys.stdin.readline()[:size.width]

        yield MapData(size, map)
        sys.stdin.readline()  # Empty line


for map in readMaps():
    map.solve()
    map.printMap()
    print()  # Empty line

print("0 0", end='')

File: shovelling/shovelling.cs
// https://open.kattis.com/problems/shovelling

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace kattis
{
    public class Program
    {
        private const string Friends = "ABCD";
        private const char Blocked = '#';
        private const char Snow = 'o';
        private const char Cleared = '.';
        private const char Shovelled = Cleared;

        private static void Main(string[] args)
        {
            while (ReadTestCase() is { } test)
            {
                var solution = Solve(test);

                PrintMap(test.Map, p =>
                {
                    if (test.Map[p.X, p.Y] == Snow && solution.Contains(p))
                        return Shovelled;
                    else
                        return test.Map[p.X, p.Y];
                });
            }

            Console.WriteLine("0 0");
        }

        private static TestCase? ReadTestCase()
        {
            string[] dim = Console.ReadLine()!.Split();
            int width = int.Parse(dim[0]);
            int height = int.Parse(dim[1]);

            if (width == 0 && height == 0)
            {
                return null;
            }

            var map = new char[width, height];
            Point[] friends = new Point[4];

            for (int y = 0; y < height; y++)
            {
                string line = Console.ReadLine()!;
                for (int x = 0; x < width; x++)
                {
                    map[x, y] = line[x];

                    if (Friends.IndexOf(line[x]) is { } p and >= 0)
                        friends[p] = new(x, y);
                }
            }

            // Read expected empty line
            Console.ReadLine();

            return new TestCase(map, height, width, friends);
        }

        private static ICollection<Point> Solve(TestCase test)
        {
            PathNode?[][,] dists = Enumerable.Range(0, 4)
                .Select(i => CalculateDistances(test.Map, test.Friends[i]))
                .ToArray();

            int bestCaseCost = 1_000_000;
            List<Point> bestCase = new List<Point>();
            HashSet<Point> current = new HashSet<Point>();

            for (int i = 0; i < test.Size; i++)
            {
                var p1 = test.GetPoint(i);
                if (test.Map[p1.X, p1.Y] == Blocked)
                    continue;

                if (dists.First()[p1.X, p1.Y] is null)
                    continue;

                var p1Distances = CalculateDistances(test.Map, p1);

                for (int j = i; j < test.Size; j++)
                {
                    var p2 = test.GetPoint(j);
                    if (test.Map[p2.X, p2.Y] == Blocked)
                        continue;

                    if (dists.First()[p2.X, p2.Y] is null)
                        continue;

                    bool stillPossible = true;

                    current.Clear();
                    foreach (var friendDistance in dists)
                    {
                        var p1Dist = friendDistance[p1.X, p1.Y];
                        var p2Dist = friendDistance[p2.X, p2.Y];

                        if (p1Dist is null || p2Dist is null)
                        {
                            stillPossible = false;
                            continue;
                        }

                        PathNode? pDist = p1Dist.Cost < p2Dist.Cost ? p1Dist : p2Dist;
                        foreach (var node in pDist.GetPath())
                        {
                            if (test.Map[node.X, node.Y] == Snow)
                            {
                                current.Add(node);
                            }
                            if (current.Count > bestCaseCost)
                            {
                                stillPossible = false;
                                break;
                            }
                        }

                        if (!stillPossible)
                            break;
                    }

                    if (!stillPossible)
                        continue;

                    // Add path between Steiner points
                    if (i != j)
                    {
                        var betweenSteinerPoints = p1Distances[p2.X, p2.Y];
                        if (betweenSteinerPoints is null)
                        {
                            stillPossible = false;
                            continue;
                        }
                        foreach (var node in betweenSteinerPoints.GetPath())
                        {
                            if (test.Map[node.X, node.Y] == Snow)
                            {
                                current.Add(node);
                            }
                        }
                    }

                    if (current.Count < bestCaseCost)
                    {
                        bestCase.Clear();
                        bestCase.AddRange(current);
                        bestCaseCost = current.Count;
                    }
                }
            }

            return bestCase;
        }

        private static void PrintMap<T, U>(T[,] map, Func<Point, U> projection)
        {
            StringBuilder output = new StringBuilder(1024);

            output.AppendFormat("{0} {1}", map.GetLength(0), map.GetLength(1)).AppendLine();

            for (int y = 0; y < map.GetLength(1); y++)
            {
                for (int x = 0; x < map.GetLength(0); x++)
                {
                    output.Append(projection(new Point(x, y)));
                }
                output.AppendLine();
            }

            output.AppendLine();
            Console.Write(output);
        }

        private static PathNode?[,] CalculateDistances(char[,] map, Point start)
        {
            PathNode?[,] result = new PathNode[map.GetLength(0), map.GetLength(1)];
            var startCost = 0;
            if (map[start.X, start.Y] == Snow)
                startCost = 1;

            result[start.X, start.Y] = new PathNode(start, startCost, null);

            Queue<Point> q = new Queue<Point>([start]);
            while (q.TryDequeue(out var p))
            {
                var currentPath = result[p.X, p.Y]!;

                Point[] nexts = [p.Left(), p.Right(), p.Up(), p.Down()];
                foreach (var next in nexts.Where(x => x.InMap(map)))
                {
                    var c = map[next.X, next.Y];
                    if (c == Blocked)
                        continue;

                    int newCost = currentPath.Cost;
                    if (c == Snow)
                        newCost += 1;
                    if (result[next.X, next.Y] is { } last && last.Cost <= newCost)
                        continue;

                    result[next.X, next.Y] = new PathNode(next, newCost, currentPath);
                    q.Enqueue(next);
                }
            }

            return result;
        }
    }

    internal record TestCase(char[,] Map, int Height, int Width, Point[] Friends)
    {
        public int Size { get; } = Height * Width;
        public int GetIndex(Point point) => point.Y * Width + point.X;
        public Point GetPoint(int index) => new Point(index % Width, index / Width);
    }
    internal record struct Point(int X, int Y)
    {
        public bool InMap(char[,] map) => X >= 0 && X < map.GetLength(0) && Y >= 0 && Y < map.GetLength(1);
        public Point Left() => new Point(X - 1, Y);
        public Point Right() => new Point(X + 1, Y);
        public Point Up() => new Point(X, Y - 1);
        public Point Down() => new Point(X, Y + 1);
        public override string ToString()
        {
            return $"{X}x{Y}";
        }
    }
    internal record class PathNode(Point Point, int Cost, PathNode? Previous)
    {
        public IEnumerable<Point> GetPath()
        {
            PathNode? node = this;
            while (node != null)
            {
                yield return node.Point;
                node = node.Previous;
            }
        }
    }
}